docs/proposals/10-shoot-additional-container-runtimes.md

```diff
@@ -0,0 +1,146 @@
+
+Since each operating system distribution has different methods of installing software (apt on ubuntu, zypper on SuSE, Torcx on CoreOS/Flatcar), we will enhance the operating system extensions to add a standard `installSoftware` script which can abstract the mechanics of installing new software on a node. The new runtime extensions will use this new functionality to install the relevant runtimes to the worker nodes.
+
+## Design Details
+
+1. Change the nodes container runtime to work with CRI and containerd by default.
+    1. In order to configure each worker machine in the cluster to work with CRI, the following configurations should be done:
+        1. Add kubelet execution flags:
+            1. --container-runtime=remote
+            2. --container-runtime-endpoint=unix:///.../containerd.socket
+        2. Make sure that containerd configuration file exist in path /etc/containerd/config.toml and it is configured to enable the CRI plugin (In the containerd config.toml "cri" plugin must not be disabled).
+        
+    2. Containerd socket path and containerd configurations are different for each OS. To make sure the configurations above are set in each worker machine, each OS extension would be responsible to configure them during the reconciliation of the 
+       OperatingSystemConfig when it creates the cloud init.
+        1. os-ubuntu - 
+            1. Add the following flags to the kubelet execution command:
+                1. --container-runtime=remote
+                2. --container-runtime-endpoint=unix:///run/containerd/containerd.sock
+            2. Create a containerd configuration file: /etc/containerd/config.toml based on the current default configuration (containerd config default).
+        2. os-coreos - 
+            1. Add the following flags to the kubelet execution command:
+               1. --container-runtime=remote
+               2. --container-runtime-endpoint=unix:///var/run/docker/libcontainerd/docker-containerd.socket
+            2. Create a containerd configuration file: /etc/containerd/config.toml based on the default configuration.
+            3. Override the default containerd.service unit. Create a new unit in: /etc/systemd/system/containerd.service unit that will start the containerd service with a configuration file path: /etc/containerd/config.toml.
+        3. os-suse-jeos - 
+           1. Add the following flags to the kubelet execution command:
+              1. --container-runtime=remote
+              2. --container-runtime-endpoint=unix:///var/run/docker/containerd/docker-containerd.sock 
+           2. Create a containerd configuration file: /etc/containerd/config.toml based on the default configuration.
+        
```

Comment:
Can you elaborate what this change will do for existing clusters running today? Let's say you restart the kubelet with these two flags - what happens with the already running containers on the Docker engine?

Reply:
We assumed that changing kubelet configuration dumping the current host and starting a new one with the update configuration, Similar to updating the OS/K8S version.
This should force rolling update of all the workers.

Reply:
Hm, this does not happen automatically. The function that triggers a rolling update is here: https://github.com/gardener/gardener-extensions/blob/1d4c446da85cb575796678fdf5ec089a279f7af1/pkg/controller/worker/machines.go#L93-L124

So, you would have to adapt the `Worker` controllers (which are calling these functions when generating the `MachineClass`es and `MachineDeployment`s, e.g. [here](https://github.com/gardener/gardener-extension-provider-aws/blob/master/pkg/controller/worker/machines.go#L111-L114)). They would have to include information about the container runtime so that a different hash value is computed (which triggers a rolling update, eventually).

However, you have to ensure that this doesn't happen without the end-user explictly triggering it. We may not trigger a rolling update for existing clusters just because we change these internal details here. This means that you have to somehow expose in the `Shoot` manifest whether containerd shall be used or not. Only if the user explicitly enables it you may change the worker pool hash. For new clusters, you can use containerd by default, I guess.

Can you add this part to the GEP? I think afterwards we can merge it. I'm not sure about the concrete implementation yet, though, these details can be discussed on the relevant PRs.
Also, please squash the commits.

Reply:
We will update the GEP to the following:

Upgrading Gardener to a version including the move to ContainerD will NOT force a rolling upgrade on existing clusters. Any cluster that is up, will continue to be up and will not be affected until the next scheduled rolling upgrade/hibernation operation. When such a cluster is resumed or a rolling upgrade occurs for any of the standard reason, the nodes will be recreated with ContainerD set as the default container management engine.

New clusters created with the upgraded Gardener will default to ContainerD.

Since simply moving from the Dockershim to the CRI endpoint doesn't change any user-visible functionality in Kubernetes, the suggested behavior should be sufficient.

Reply:
Squashing the commits will be done when merging (There is an option to merge the code with one commit)

Reply:
it seems auto squashing did not work :cry: